lib/private/TaskProcessing/Manager.php

```diff
@@ -439,35 +511,64 @@ private function _getTaskTypes(): array {
 
 	/**
 	 * @param ShapeDescriptor[] $spec
+	 * @param array<array-key, string|numeric> $defaults
+	 * @param array<array-key, ShapeEnumValue[]> $enumValues
 	 * @param array $io
+	 * @param bool $optional
 	 * @return void
 	 * @throws ValidationException
 	 */
-	private function validateInput(array $spec, array $io, bool $optional = false): void {
+	private static function validateInput(array $spec, array $defaults, array $enumValues, array $io, bool $optional = false): void {
 		foreach ($spec as $key => $descriptor) {
 			$type = $descriptor->getShapeType();
 			if (!isset($io[$key])) {
 				if ($optional) {
 					continue;
 				}
+				if (isset($defaults[$key])) {
+					if (EShapeType::getScalarType($type) !== $type) {
+						throw new ValidationException('Provider tried to set a default value for a non-scalar slot');
+					}
+					if (EShapeType::isFileType($type)) {
+						throw new ValidationException('Provider tried to set a default value for a slot that is not text or number');
+					}
+					$type->validateInput($defaults[$key]);
+					continue;
+				}
 				throw new ValidationException('Missing key: "' . $key . '"');
 			}
 			try {
 				$type->validateInput($io[$key]);
+				if (isset($enumValues[$key])) {
+					$type->validateEnum($io[$key], $enumValues[$key]);
+				}
 			} catch (ValidationException $e) {
 				throw new ValidationException('Failed to validate input key "' . $key . '": ' . $e->getMessage());
 			}
 		}
 	}
 
+	/**
+	 * Takes task input data and replaces fileIds with File objects
+	 *
+	 * @param array<array-key, list<numeric|string>|numeric|string> $input
+	 * @param array<array-key, numeric|string> ...$defaultSpecs the specs
+	 * @return array<array-key, list<numeric|string>|numeric|string>
+	 */
+	public function fillInputDefaults(array $input, ...$defaultSpecs): array {
+		$spec = array_reduce($defaultSpecs, fn ($carry, $spec) => $carry + $spec, []);
+		return $spec + $input;
```

Comment:
```suggestion
		$spec = array_reduce($defaultSpecs, fn ($carry, $spec) => array_merge($carry, $spec), []);
		return array_merge($spec, $input);
```

This always writes default values even when a value is set in the submitted input.